README.md
# nuxt-bulma-breakpoints

> Resize observer breakpoints with Nuxt.js module. Bulma edition

## Setup

1. Add `nuxt-bulma-breakpoints` dependency to your project

```bash
yarn add nuxt-bulma-breakpoints # or npm install nuxt-breakpoints
```

2. Add `nuxt-bulma-breakpoints` to the `modules` section of `nuxt.config.js`

```js
{
  modules: [
    // Simple usage
    'nuxt-breakpoints',

    // With options
    ['nuxt-breakpoints', { /* module options */ }]
  ]

  // Another way to use options
  breakpoints: {
    // default options
    tablet: 769,
    desktop: 1024,
    widescreen: 1216,
    fullhd: 1408,
    options: {
      polyfill: true,
      throttle: 200
    }
  }
}
```
## Usage
```js
// components.vue
export default {
  computed: {
    isMobile() {
      return this.$breakpoints.mobile
    },
    untilWidescreen(){
      return this.$breakpoints.untilWidescreen
    }
    current() {
      return this.$breakpoints.current
    }
  }
}
```

## Breakpoints and helpers

| Helper          | Note                                                     |
|-----------------|----------------------------------------------------------|
| mobile          | width < 769px (tablet)                                   |
| tablet          | width >= 769px (tablet)                                  |
| tabletOnly      | width >= 769px (tablet) && width < 1024 (desktop)        |
| touch           | width < 1024px (desktop)                                 |
| desktop         | width >= 1024px (desktop)                                |
| desktopOnly     | width >= 1024px (desktop) && width < 1216px (widescreen) |
| untilWidescreen | width < 1216px (widescreen)                              |
| widescreen      | width >= 1216px (widescreen)                             |
| widescreenOnly  | width >= 1216px (widescreen) && width < 1408px (fullhd)  |
| untilFullhd     | width < 1408px (fullhd)                                  |
| fullhd          | width >= 1408px (fullhd)                                 |

## Options

| Property   | Type   | Default                           |
|------------|--------|-----------------------------------|
| mobile     | none   | none                              |
| tablet     | number | 769                               |
| desktop    | number | 1024                              |
| widescreen | number | 1216                              |
| fullhd     | number | 1408                              |
| options    | object | { polyfill: true, throttle: 200 } |

**Note**: `polyfill` default by true, which means auto-import `resize-observer-polyfill` when the browser doesn't support ResizeObserver more information below, `throttle` will slow down when Window has resizing trigger speed.

## Development

1. Clone this repository
2. Install dependencies using `yarn install` or `npm install`
3. Start development server using `yarn dev` or `npm run dev`

## Ref
[Nuxt.js](https://nuxtjs.org)

[MDN - ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver)

[ResizeObserver Polyfill](https://github.com/que-etc/resize-observer-polyfill)

## License

[MIT License](./LICENSE)

Copyright (c) Steven Ho <shockshocksoul@gmail.com>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.1.0](https://github.com/Rosendito/nuxt-bulma-breakpoints/compare/v1.1.9...v1.1.0) (2020-07-13)


### Features

* **core:** Change breakpoint names and value according to bulma ([704e82b](https://github.com/Rosendito/nuxt-bulma-breakpoints/commit/704e82bdaae5cb655aa13bf8c4e0ef6311bcf8ad))
* **docs:** Update readme with new value options ([24c4768](https://github.com/Rosendito/nuxt-bulma-breakpoints/commit/24c47682936732795162d3fd0c4bd1a004b10a1e))
* **example:** Update example ([0c4587e](https://github.com/Rosendito/nuxt-bulma-breakpoints/commit/0c4587e0b8c417b66e3a043d63d4af9ba0dfd0b7))
* **init:** Iniciar repo de nuxt-bulma-breakpoints ([c053f18](https://github.com/Rosendito/nuxt-bulma-breakpoints/commit/c053f184c4f22b0316e6e12cffaa3707898ccb2a))
* **package:** Update package name ([8afeb71](https://github.com/Rosendito/nuxt-bulma-breakpoints/commit/8afeb71fb5596378a44d52f67e125312564f023f))

### [1.1.9](https://github.com/steven0811/nuxt-breakpoints/compare/v1.1.8...v1.1.9) (2019-12-01)

### [1.1.8](https://github.com/steven0811/nuxt-breakpoints/compare/v1.1.7...v1.1.8) (2019-12-01)

### [1.1.6](https://github.com/steven0811/nuxt-breakpoints/compare/v1.1.5...v1.1.6) (2019-10-20)

### [1.1.5](https://github.com/steven0811/nuxt-breakpoints/compare/v1.1.4...v1.1.5) (2019-10-19)

### [1.1.4](https://github.com/steven0811/nuxt-breakpoints/compare/v1.1.3...v1.1.4) (2019-10-19)

### [1.1.3](https://github.com/steven0811/nuxt-breakpoints/compare/v1.1.2...v1.1.3) (2019-10-19)

### [1.1.2](https://github.com/steven0811/nuxt-breakpoints/compare/v1.1.1...v1.1.2) (2019-10-19)

### [1.1.1](https://github.com/steven0811/nuxt-breakpoints/compare/v1.0.8...v1.1.1) (2019-10-19)

### [1.0.8](https://github.com/steven0811/nuxt-breakpoints/compare/v1.0.7...v1.0.8) (2019-10-19)

### [1.0.7](https://github.com/steven0811/nuxt-breakpoints/compare/v1.0.6...v1.0.7) (2019-10-19)

### [1.0.6](https://github.com/steven0811/nuxt-breakpoints/compare/v1.0.5...v1.0.6) (2019-10-19)

### [1.0.5](https://github.com/steven0811/nuxt-breakpoints/compare/v1.0.4...v1.0.5) (2019-10-19)

### [1.0.4](https://github.com/steven0811/nuxt-breakpoints/compare/v1.0.3...v1.0.4) (2019-10-19)

### [1.0.3](https://github.com/steven0811/nuxt-breakpoints/compare/v1.0.2...v1.0.3) (2019-10-19)

### [1.0.2](https://github.com/steven0811/nuxt-breakpoints/compare/v1.0.1...v1.0.2) (2019-10-19)

### 1.0.1 (2019-10-19)


lib/plugin.js
import Vue from 'vue'
import throttle from 'lodash/throttle'
import { transform } from 'lodash'
const extend = (app, mixin) => {
  if (!app.mixins) {
    app.mixins = []
  }
  app.mixins.push(mixin)
}


const defaultOptions = {
  tablet: 769,
  desktop: 1024,
  widescreen: 1216,
  fullhd: 1408,
  options: {
    polyfill: true,
    throttle: 200
  }
}

const defaultBreakpoints = {
  current: 'mobile',

  mobile: true,
  tablet: false,
  tabletOnly: false,
  touch: false,
  desktop: false,
  desktopOnly: false,
  untilWidescreen: true,
  widescreen: false,
  widescreenOnly: false,
  untilFullhd: true,
  fullhd: false,

  width: 0,
  height: 0
}

const conditions = {
  mobile: (width, breakpoints) => width < breakpoints.tablet,
  tablet: (width, breakpoints) => width >= breakpoints.tablet,
  tabletOnly: (width, breakpoints) => width >= breakpoints.tablet && width < breakpoints.desktop,
  touch: (width, breakpoints) => width < breakpoints.desktop,
  desktop: (width, breakpoints) => width >= breakpoints.desktop,
  desktopOnly: (width, breakpoints) => width >= breakpoints.desktop && width < breakpoints.widescreen,
  untilWidescreen: (width, breakpoints) => width < breakpoints.widescreen,
  widescreen: (width, breakpoints) => width >= breakpoints.widescreen,
  widescreenOnly: (width, breakpoints) => width >= breakpoints.widescreen && width < breakpoints.fullhd,
  untilFullhd: (width, breakpoints) => width < breakpoints.fullhd,
  fullhd: (width, breakpoints) => width >= breakpoints.fullhd,
}

const transformBreakpoints = (breakpoints, { width, height }, options) => {
  const originalBreakpoints = [
    'mobile',
    'tablet',
    'desktop',
    'widescreen',
    'fullhd'
  ]
  const conditionsKey = Object.keys(conditions)
  let transformData = {
    current: 'mobile'
  }

  conditionsKey.forEach(breakpoint => {
    transformData[breakpoint] = conditions[breakpoint](width, options)

    if (
      originalBreakpoints.includes(breakpoint) &&
      conditions[breakpoint](width, options)
    )
      transformData.current = breakpoint
  })

  transformData.width = ~~width
  transformData.height = ~~height

  Object.assign(breakpoints, transformData)
}

export default async (ctx, inject) => {
  const options = { ...defaultOptions, ...<%= JSON.stringify(options, null, 2) %>}


  const breakpoints = Vue.observable(defaultBreakpoints)

  if (process.server) {
    ctx.$breakpoints = breakpoints
    inject('breakpoints', breakpoints)
    return
  }

  const needPolyfill =
    !Object.prototype.hasOwnProperty.call(window, 'ResizeObserver') &&
    options.options.polyfill



  if (needPolyfill) {
    const ResizeObserver = await import('resize-observer-polyfill')

    Object.defineProperty(window, 'ResizeObserver', {
      value: ResizeObserver.default,
      writable: false
    })
  }

  const bodyElem = document.querySelector('body')

  const resizeObserver = new ResizeObserver(
    throttle((entries) => {
      const [{ contentRect }] = entries
      transformBreakpoints(breakpoints, contentRect, options)
    }, +options.options.throttle, { trailing: true, leading: false })
  )

  extend(ctx.app, {
    beforeDestroy () {
      resizeObserver.unobserve(bodyElem)
    }
  })

  resizeObserver.observe(bodyElem)

  ctx.$breakpoints = breakpoints
  inject('breakpoints', breakpoints)
}
